wordle-web/scripts/validWordsUtils.ts

```diff
@@ -0,0 +1,168 @@
+      let lettersIndex = character.charCodeAt(0) - 65;
+      let letterHelper = this.letters[lettersIndex];
+      letterHelper.runningTotalDiscovered++;
+      // Also checks if correct letters are found in the word
+      if (
+        (this.partialWord[index] !== undefined &&
+          this.partialWord[index] !== character) ||
+        !letterHelper.isValidForIndex[index]
+      ) {
+        return false;
+      }
+    }
+    // Makes sure there's at least however many letters already discovered (through misplaced and correct letters) in current word list word
+    for (
+      let lettersIndex = 0;
+      lettersIndex < this.letters.length;
+      lettersIndex++
+    ) {
+      let letter = this.letters[lettersIndex];
+      if (letter.runningTotalDiscovered < letter.numberDiscoveredInWord) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  // Set up valid for index, number found, number discovered based on current guesses
+  setup() {
+    for (
+      let guessesIndex = 0;
+      guessesIndex < this.game.guessIndex;
+      guessesIndex++
+    ) {
+      let guess = this.game.guesses[guessesIndex];
+      this.setRunningTotalForAllLettersToZero();
+      if (!guess.isFilled) {
+        return;
+      }
+      let lettersArray = guess.letters;
+      // Adjust letter helpers based on correct, misplaced, or wrong states
+      for (let index = 0; index < lettersArray.length; index++) {
+        let lettersIndex = lettersArray[index].char.charCodeAt(0) - 65;
+        if (lettersArray[index].state === LetterState.Correct) {
+          this.partialWord[index] = lettersArray[index].char;
+          this.letters[lettersIndex].numberFound++;
+          this.letters[lettersIndex].runningTotalDiscovered++;
+        } else if (lettersArray[index].state === LetterState.Misplaced) {
+          this.letters[lettersIndex].isValidForIndex[index] = false;
+          this.letters[lettersIndex].runningTotalDiscovered++;
+        } else if (lettersArray[index].state === LetterState.Wrong) {
+          let letterHelper = this.letters[lettersIndex];
+          if (letterHelper.numberFound === letterHelper.numberInSecretWord) {
+            this.letters[lettersIndex].isValidForIndex = Array.from(
+              { length: this.partialWord.length },
+              (v, k) => {
+                return (
+                  this.partialWord[k] === this.letters[lettersIndex].character
+                );
+              }
+            );
+          } else {
+            this.letters[lettersIndex].isValidForIndex[index] = false;
+          }
+        }
+      }
+      // Update total discovered in word
+      for (
+        let lettersIndex = 0;
+        lettersIndex < this.letters.length;
+        lettersIndex++
+      ) {
+        this.letters[lettersIndex].numberDiscoveredInWord =
+          this.letters[lettersIndex].runningTotalDiscovered;
+        this.letters[lettersIndex].runningTotalDiscovered = 0;
+      }
+    }
+  }
+}
+

```

Comment:
I like the implementation for this, it took me a while to understand it but seems super robust.

Reply:
Yeah, I'm not sure if it's the prettiest code, but it works.